### PR TITLE
Backend: `get_aliased_box` should not create boxes

### PR DESCRIPTION
**Pull request by nomeata**
The semantics of the initializer of a mutable field in `NewObjE` is that
it should alias the value; this is needed for async/await.

To be aliased, that variable must be put in the heap, as a `HeapInd`.
Previously, if that was not the case, we’d simply create a new box. But
that breaks the semantics! (I don’t think this could be observed in
source, though).

Better if `get_aliased_box` fails if it doesn’t find the value allocated
on the heap. And to make sure it always is on the heap, we have to treat
the initializer in `NewObjE` a bit like a mutable variable captured by a
lambda. And then it seems to work.

### CODE REVIEW

**Comment by nomeata** [comment]
I couldn't run the tests locally (something wrong with wasmtime). Will have to investigate the CI failures.

File: src/codegen/compile.ml [hunk 1]
```diff
@@ -6780,14 +6780,14 @@ module Var = struct
...
+     In the IR, mutable fields of objects are pre-allocated as MutBox objects,
+     to allow the async/await.
+     So we expect the variable to be in a HeapInd (pointer to MutBox on the heap),
+     and we use the pointer.
   *)
   let get_aliased_box env ae var = match VarEnv.lookup_var ae var with
     | Some (HeapInd i) -> G.i (LocalGet (nr i))
-    | Some (HeapStatic _) -> assert false (* we never do this on the toplevel *)
-    | _  -> Tagged.obj env Tagged.ObjInd [ get_val_vanilla env ae var ]
+    | _ -> assert false
```
[the remark this answers is not in the document]
Comment (ggreif) [reply src/codegen/compile.ml:6790]: And there I have a regression
``` diff
show: [tc] [run] [run-ir] [run-low] [comp] [valid] [wasm-run]
--- show.wasm-run (expected)
+++ show.wasm-run (actual)
@@ -48,7 +48,7 @@
 [1, 2, 3]
 [var]
 [var 1, 2, 3]
-{bar = true; foo = +42}
+{bar = false; foo = +42}
 #foo
 #bar(1, 2)
 #bar({})
```

**Comment by nomeata** [comment]
Ok, this seems to work now. 

@ggreif, this simplifies your work, right?

**Comment by nomeata** [comment]
Oh, I think this can be improved, and the Code Smell be fixed:

If we have a mutable variable on the top level that’s used in a record, then we can still allocate it statically, and put the pointer to that static location into the record!

**Comment by nomeata** [comment]
Ok, this is prettier

File: src/ir_def/freevars.ml [hunk 2]
```diff
@@ -95,7 +96,11 @@ let args as_ : fd = union_binders arg as_
 
 let id i = M.singleton i {captured = false; eager = true}
 
-let fields fs = unions (fun f -> id f.it.var) fs
+(* The mutable fields of an IR object behave a bit like a lambda, in that they capture mutable
+boxes by reference. So set captured = true for them. *)
+let fields fs = unions (fun f ->
+  M.singleton f.it.var {captured = Type.is_mut f.note; eager = true}
```
[the remark this answers is not in the document]
Comment (ggreif) [reply src/ir_def/freevars.ml:102]: Nope, it is defined in up here.

**Review by ggreif — APPROVED** [review]
Short and sweet, I like it!

**Comment by ggreif** [comment]
> and it will work out nicely.

I suspect that is not enough. A mutable cell is still being created (with bogus content: `0`, then the original mutbox pointer)

``` Motoko
let e = { e = 42 in c };  // base c has a `c` field
```

``` wasm
    i32.const 2
    call $alloc_words            -- mutbox
    local.tee $heap_object
    i32.const 9
    i32.store offset=1
    local.get $heap_object
    i32.const 0                    -- WUT?
    i32.store offset=5
    local.get $heap_object
    local.set $$c/6
    local.get $c
    local.tee $$base/17
    i32.const 99
    call $obj_idx<0>
    i32.load offset=1        -- mutbox pointer from `c`
    local.set $new_val.1
    local.get $$c/6
    local.get $new_val.1
    i32.store offset=5        -- another store into the mutbox
    i32.const 5
    call $alloc_words         -- alloc of `e`
    local.tee $obj
    i32.const 1
    i32.store offset=1
    local.get $obj
    i32.const 2
    i32.store offset=5
    local.get $obj
    i32.const 65747
    i32.store offset=9
    local.get $obj
    i32.const 84            -- `42`
    i32.store offset=17  -- store of field `e`
    local.get $obj
    local.get $$c/6
    i32.store offset=13  --  store of field `c`
    local.get $obj
    local.set $e
```

**Pull request merged**